From e17675eccbab0e41736a50aba62ac540174da15d Mon Sep 17 00:00:00 2001
From: Giovanni Mascellani <gmascellani@codeweavers.com>
Date: Fri, 17 Jan 2025 13:52:39 +0100
Subject: [PATCH] dxgi: Set the frame latency even when increasing the frame
 latency.

This is only partially correct, because the frame latency waitable
should be a semaphore instead. However, it's a step in the right
direction, in preparation to make the waitable an actual semaphore.
---
 dlls/dxgi/swapchain.c  | 9 +++++++++
 dlls/dxgi/tests/dxgi.c | 2 +-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/dlls/dxgi/swapchain.c b/dlls/dxgi/swapchain.c
index b1ce578e8be..05ffe628d95 100644
--- a/dlls/dxgi/swapchain.c
+++ b/dlls/dxgi/swapchain.c
@@ -2886,6 +2886,15 @@ static HRESULT STDMETHODCALLTYPE d3d12_swapchain_SetMaximumFrameLatency(IDXGISwa
         return DXGI_ERROR_INVALID_CALL;
     }
 
+    if (max_latency > swapchain->frame_latency)
+    {
+        if (!SetEvent(swapchain->frame_latency_event))
+        {
+            ERR("Failed to set event, last error %lu.\n", GetLastError());
+            return HRESULT_FROM_WIN32(GetLastError());
+        }
+    }
+
     swapchain->frame_latency = max_latency;
     return S_OK;
 }
diff --git a/dlls/dxgi/tests/dxgi.c b/dlls/dxgi/tests/dxgi.c
index 912bcd8fa61..7cf52d626da 100644
--- a/dlls/dxgi/tests/dxgi.c
+++ b/dlls/dxgi/tests/dxgi.c
@@ -7143,7 +7143,7 @@ static void test_frame_latency_event(IUnknown *device, BOOL is_d3d12)
     ok(frame_latency == 3, "Got unexpected frame latency %#x.\n", frame_latency);
 
     wait_result = WaitForSingleObject(semaphore, 0);
-    todo_wine ok(!wait_result, "Got unexpected wait result %#lx.\n", wait_result);
+    ok(!wait_result, "Got unexpected wait result %#lx.\n", wait_result);
     wait_result = WaitForSingleObject(semaphore, 0);
     todo_wine ok(!wait_result, "Got unexpected wait result %#lx.\n", wait_result);
     wait_result = WaitForSingleObject(semaphore, 100);
-- 
GitLab

